### DISCUSSION TITLE
Able to Callable(BuiltInObject) especially Callable(Signal)

### DISCUSSION THREAD

**Discussion by comods** [godotengine/godot-proposals#4107] [Scripting]
A Callable references a method of an object for callbacks and plugins.

Callables can execute methods of non-built-in object types **but not** methods of native built-in types, especially **Callable(Signal)**, because built-ins do not inherit from the class Object.

Possible fixes AA, BB, etc are given below for voting and attaching feedback.
None of these fixes work with Godot 4 alpha 3.
To promote fixing this issue, vote on this post, and each fix below that you would welcome.

**Comment by comods** [comment]
Fix AA:

AA is for reducing restrictions on argument type.

Callable(myObject.mySignal, "emit") # requires allowing argument to be a built-in.

**Comment by comods** [comment]
Fix BB:

BB is for reducing restrictions on bind (this fixes for Signals on non-built-in Objects).

Callable(myObject, "emit_signal").bind("mySignal")
    # requires bind() from left with rbind() from right, instead of bind() from right.

**Comment by comods** [comment]
Fix CC:

CC is for arguments to allow Signals (this only fixes for Signals and not other built-ins).

Callable(myObject.mySignal) # requires allowing first argument to be a Signal.
Callable(myObject, "mySignal") # requires allowing second argument to be a SignalName.

**Comment by comods** [comment]
Fix DD:

DD is for allowing casting Signals into Objects (this only fixes for Signals and not other built-ins).

Callable(myObject.mySignal as Object, "emit") # requires allowing Signals to be cast as Objects.
\# As Object is at your own risk, so don't access non-existent methods or properties.

**Comment by azur-wolve** [comment]
@RefinedSoftwareLLC can't understand your posts enough well because lack of context;

What AA, BB, CC, [...] means?

(also better to group all your comments with the main post)

**Reply by Zireael07** [reply]
Several ways of achieving the goal

**Reply by comods** [reply]
They are Tags to refer to each presented solution during discussion and to allow people to up-vote and reply to individual fixes.

**Reply by comods** [reply]
@KiwwiPity I have added detail to the main post to address your concerns.

**Comment by dalexeev** [comment]
This is implemented for future 4.3:

* godotengine/godot#82264
* godotengine/godot#88948
